New DSC resource to install tools from dotnet globally

Gijsreyn opened the pull request:
<!-- To check a checkbox place an "x" between the brackets. e.g: [x] -->

- [x] I have signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-dsc).
- [ ] This pull request is related to an issue.

-----

This pull request introduces a new DSC resource. It allows users to install .NET tools globally.

Gijsreyn commented:
@denelon Hi Demitrius, I had gone through the `CONTRIBUTING.md`, but I was unsure if I had to commit the spec-template file with this PR. Do you want me to include it?

denelon commented:
Yes please. Any additional documentation will be helpful as users become informed about the new resource.

denelon commented:
It might also be interesting to get you connected to some of the dotnet folks to discuss making dotnet.exe DSC v3 capable. I saw you're writing an e-book on DSC v3 in your GitHub profile.

Gijsreyn commented:
> It might also be interesting to get you connected to some of the dotnet folks to discuss making dotnet.exe DSC v3 capable. I saw you're writing an e-book on DSC v3 in your GitHub profile.

Would love that Demitrius. Great challenge to look at. Indeed I am. Currently storing the content from preview.10 and asked Gael to have a review/foreword. Then probably when dsc gets GA, I'll put it fully live.

Gijsreyn commented:
> We have been discussing a DSC Resource for dotnet recently. I would have thought about using a module name like "Microsoft.dotnet.Dsc" or something like that. I just don't have enough domain knowledge on dotnet to make the best recommendation.

I would recon a naming change. It was just as first thought that popped up.

denelon commented:
Related to:
* https://github.com/microsoft/winget-dsc/issues/65

MichaelSimons commented on resources/Microsoft.NET.SDK.ToolInstaller/Microsoft.NET.SDK.ToolInstaller.psd1:
```diff
@@ -0,0 +1,24 @@
+    RootModule = 'Microsoft.NET.SDK.ToolInstaller.psm1'
+    ModuleVersion = '0.1.0'
+    GUID = '2e883e78-1d91-4d08-9fc1-2a968e31009d'
+    Author = 'Microsoft Corporation'
```
@denelon - Is there an ownership concept with DSCs similar to Winget packages?  How do you ensure the validity of the author and that they have sole ownership?

denelon replied:
Since this is an OSS repository, we'd consider PRs from the community, but since this is a Microsoft "package" the resource would require our approval to make sure we're good with the implementation. Anything published to the PowerShell gallery from here would need that validation before we would sign and publish the module.

marcpopMSFT commented on resources/Microsoft.NET.SDK.ToolInstaller/Microsoft.NET.SDK.ToolInstaller.psm1:
```diff
@@ -0,0 +1,297 @@
+
+#region Functions
+function Get-DotNetPath {
+    if ($IsWindows) {
```
Are you only supporting default install paths for each of these? Customers can add dotnet to the path themselves or do an app-local install like we do in our own repos.

Gijsreyn replied:
It's still as TODO task on the class. I will try to add this option also.

Gijsreyn replied:
It has been added

marcpopMSFT commented on resources/Microsoft.NET.SDK.ToolInstaller/Microsoft.NET.SDK.ToolInstaller.psm1:
```diff
@@ -0,0 +1,297 @@
+#region Functions
+function Get-DotNetPath {
+    if ($IsWindows) {
+        $dotNetPath = "$env:ProgramFiles\dotnet\dotnet.exe"
```
Isn't this for VSCode? If so, I don't believe that supports x86 so you probably don't need that. I assume you also wont' need dotnet\x64.

marcpopMSFT commented on resources/Microsoft.NET.SDK.ToolInstaller/Microsoft.NET.SDK.ToolInstaller.psm1:
```diff
@@ -0,0 +1,297 @@
+    return $dotNetPath
+}
+
+# TODO: when https://github.com/dotnet/sdk/pull/37394 is documented and version is released with option simple use --format=JSON
```
The linked PR is fixed in 9 it looks like. Were you waiting for 9 to be the default to replace the below code?

Gijsreyn replied:
You are right, I forgot I couldn't install the latest preview from the official website. Everytime when I tried downloading, it gives me 0 bytes. When I used `winget install Microsoft.DotNet.SDK.Preview`, it did. I guess it installs from a different location.

Eventually, I guess we can move the logic to this:

![image](https://github.com/user-attachments/assets/119ce361-435a-40c2-ac77-30572d57e50d)

But we should also be able to support older versions of the SDK?

marcpopMSFT commented on resources/Microsoft.NET.SDK.ToolInstaller/Microsoft.NET.SDK.ToolInstaller.psm1:
```diff
@@ -0,0 +1,297 @@
+    process
+    {
+        $installArgument = Get-DotNetToolArguments @PSBoundParameters
+        $arguments = "tool install $installArgument --global --ignore-failed-sources"
```
are all of the tools you're installing supported on the current runtimes that are installed? if not, there may be issues with running those tools.

Gijsreyn replied:
I agree with your statement here, however I do think users themselves should be aware what they should install. I know the architecture option is also available to be specified. It can be added later I would say. Let me know your thoughts.

marcpopMSFT commented on resources/Microsoft.NET.SDK.ToolInstaller/Microsoft.NET.SDK.ToolInstaller.psm1:
```diff
@@ -0,0 +1,297 @@
+    }
+
+    Set() {
+        # TODO: validate for upgrade/update scenarios
```
what is the update model for this?

Gijsreyn replied:
Let's imagine the following scenario. The user has already installed PowerShell version 7.4.4. The user now wants version 7.4.5. The user leaves empty the version property. When the user runs the configuration, it needs to check if 7.4.4 is installed and if so, update it with `dotnet tool update`. @ryfu-msft Would you agree with such logic, or should we add a boolean parameter to upgrade?

ryfu-msft replied:
Agreed. If a package has a version that does not match the desired version, then this DSC resource should handle updating the package to match the desired state.

marcpopMSFT commented on tests/Microsoft.NET.SDK.ToolInstaller/Microsoft.NET.SDK.ToolInstaller.Tests.ps1:
```diff
@@ -0,0 +1,103 @@
+     It 'Sets desired package with prerelease version' -Skip:(!$IsWindows) {
+        $parameters = @{
+            PackageId = 'PowerShell'
+            Version = '7.2.0-preview.5'
```
why this version of powershell?

Gijsreyn replied:
If I'm not mistaken, the logic of the CLI to install packages is as followed:

1. Whenever a preview version is higher than the official version, you require the `--prerelease` parameter to be added without _version_
2. When the official version is higher than a preview version, you use the `--version` parameter to indicate you want to install a prerelease version that is older. Because you want to test for this scenario, I have included an older version of PowerShell. See list:

![image](https://github.com/user-attachments/assets/6250250c-13b1-43de-a2bd-bb03406fc4b5)

Gijsreyn commented:
@denelon In the latest commit, I have added the option for toolsPath. This allows users to choose a directory to install tools towards.

However, the latest pull request for `Microsoft.VSCode.DSC`, broke the participation on the export because we removed the static method. I asked Mikey for feedback on enabling export for class-based resources directly from [PSDesiredStateConfiguration](https://github.com/PowerShell/PSDesiredStateConfiguration/issues/123).

I have tried to incorporate as much of what Mikey sketched out in this change. It probably requires updating the `Microsoft.VSCode.DSC` also if this is the way forward.

P.S. After thinking about it, `DotNetDsc` would to it for me.

denelon commented:
> @denelon In the latest commit, I have added the option for toolsPath. This allows users to choose a directory to install tools towards.
> 
> However, the latest pull request for `Microsoft.VSCode.DSC`, broke the participation on the export because we removed the static method. I asked Mikey for feedback on enabling export for class-based resources directly from [PSDesiredStateConfiguration](https://github.com/PowerShell/PSDesiredStateConfiguration/issues/123).
> 
> I have tried to incorporate as much of what Mikey sketched out in this change. It probably requires updating the `Microsoft.VSCode.DSC` also if this is the way forward.
> 
> P.S. After thinking about it, `DotNetDsc` would to it for me.

I'd suggest trying with the experimental WinGet feature something like:

`winget configure export --Resource Microsoft.dotnet.Dsc -o dotnet.winget` to see if the settings get exported.

I know this is brutal with class-based v2 modules, but it gives us another working example and helps pave the way for DSC v3. We're investing into improvements for the export feature to capture all installed packages, Windows Settings, and any DSC v3 resources (soon to be identified in WinGet package manifests).

ryfu-msft commented on resources/Microsoft.NET.SDK.ToolInstaller/Microsoft.NET.SDK.ToolInstaller.psd1:
I would rename this entire DSC to `Microsoft.DotNet.Dsc`. It seems like everything in this module will relate to `dotnet.exe` so I feel that it needs a more generic name. It also needs to include `Dsc` as that is the standard convention.

ryfu-msft commented on resources/Microsoft.NET.SDK.ToolInstaller/Microsoft.NET.SDK.ToolInstaller.psm1:
```diff
@@ -0,0 +1,404 @@
+            $defaultDir = " "
+            if ($toolsDir)
+            {
+                # TODO: How are we going to handle if tool path was not set? Should exist be false?
```
Yes exist should be false because no packages exist in that tools directory.

ryfu-msft commented on resources/Microsoft.NET.SDK.ToolInstaller/Microsoft.NET.SDK.ToolInstaller.psm1:
```diff
@@ -0,0 +1,404 @@
+
+            # default directory
+            $defaultDir = " "
+            if ($toolsDir)
```
Does this handle a tools directory that does not exist? We may need to show an error to the user if that directory doesn't exist.

ryfu-msft commented on resources/Microsoft.NET.SDK.ToolInstaller/Microsoft.NET.SDK.ToolInstaller.psm1:
```diff
@@ -0,0 +1,404 @@
+    [DscProperty()]
+    [bool] $PreRelease = $false
+
+    [DscProperty()]
+    [string] $ToolPath
```
Do we need another dsc boolean property for `Global`

Gijsreyn replied:
I don't think so Ryan, please check latest commit as we default to global.

ryfu-msft commented on resources/Microsoft.NET.SDK.ToolInstaller/Microsoft.NET.SDK.ToolInstaller.psm1:
```diff
@@ -0,0 +1,404 @@
+                $pre = $true
+            }
+
+            # default directory
+            $defaultDir = " "
```
nit: Is this the same as `$installDirectory`?  That seems a little bit more clear to me but just my opinion.

ryfu-msft commented on resources/Microsoft.NET.SDK.ToolInstaller/Microsoft.NET.SDK.ToolInstaller.psm1:
```diff
@@ -0,0 +1,404 @@
+    }
+    #endregion NETSDKToolInstaller helper functions
+}
+#
+#endregion Classesendregion Classes
```
```suggestion
#endregion Classes
```

ryfu-msft commented on pipelines/azure-pipelines.yml:
```diff
@@ -44,6 +44,7 @@ extends:
         displayName: "Publish WinGet DSC Resources"
         templateContext:
           outputs:
+          # TODO: Add pipeline artifact for NET.SDK.ToolInstaller
```
```suggestion
          - output: pipelineArtifact
            displayName: 'Publish Pipeline Microsoft.DotNet.Dsc'
            targetPath: $(Build.SourcesDirectory)\resources\Microsoft.DotNet.Dsc\
            artifactName: Microsoft.DotNet.Dsc
```
Append this to the end of this list and we should be able to run the tests.

Gijsreyn commented:
@ryfu-msft Ryan, sorry for the many commits. I had to re-name the module and figure things out with Semantic Versioning. I have tried to prepare the test to run on the ADO agents. Inside the `.psd1` file, the `LicenseUri` and `ProjectUri` are included. This is more convenient for people to be redirected when they want to do changes. You mind giving it another look and run the pipelines? 

Cheers and have a good weekend already.

ryfu-msft commented:
/azp run

ryfu-msft commented on resources/Microsoft.DotNet.Dsc/Microsoft.DotNet.Dsc.psm1:
```diff
@@ -0,0 +1,547 @@
+        [bool]   $PreRelease,
+        [Parameter(Mandatory = $false)]
+        [string] $ToolPath,
+        [bool]   $Exist
```
Is this missing a parameter attribute?

ryfu-msft commented on resources/Microsoft.DotNet.Dsc/Microsoft.DotNet.Dsc.psm1:
```diff
@@ -0,0 +1,547 @@
+    }
+    
+    $PSBoundParameters.GetEnumerator() | ForEach-Object {
+        if ($mappingTable.ContainsKey($_.Key) -and $_.Value -ne $false)
```
Does `$.Value -ne $false` work for parameters that are not boolean parameters like $Version?

Gijsreyn replied:
Fixed with one more additional check

ryfu-msft commented on resources/Microsoft.DotNet.Dsc/Microsoft.DotNet.Dsc.psm1:
```diff
@@ -0,0 +1,547 @@
+                }
+                return $true
+            })]
+        [string] $ToolPath,
```
nit:
```suggestion
        [string] $ToolPathDirectory,
```

ryfu-msft commented on resources/Microsoft.DotNet.Dsc/Microsoft.DotNet.Dsc.psm1:
```diff
@@ -0,0 +1,547 @@
+    foreach ($package in $packages)
+    {
+        # flags to determine the existence of the package
+        $preFlag = $false
```
```suggestion
        $isPrerelease= $false
```

ryfu-msft commented on resources/Microsoft.DotNet.Dsc/Microsoft.DotNet.Dsc.psm1:
```diff
@@ -0,0 +1,547 @@
+    [string[]] $Commands
+
+    [DscProperty()]
+    [bool] $PreRelease = $false
```
nit:
```suggestion
    [bool] $Prerelease = $false
```

ryfu-msft commented on resources/Microsoft.DotNet.Dsc/Microsoft.DotNet.Dsc.psm1:
```diff
@@ -0,0 +1,547 @@
+    [bool] $PreRelease = $false
+
+    [DscProperty()]
+    [string] $ToolPath
```
```suggestion
    [string] $ToolPathDirectory
```

ryfu-msft commented on resources/Microsoft.DotNet.Dsc/Microsoft.DotNet.Dsc.psm1:
```diff
@@ -0,0 +1,547 @@
+        # get the current state
+        $currentState = [DotNetToolPackage]::InstalledPackages[$this.PackageId]
+
+        # get the properties of the object currently set
+        $properties = $this.ToHashTable()
+
+        # refresh installed packages
+        [DotNetToolPackage]::GetInstalledPackages($properties)
+
+        # current state
+        $currentState = [DotNetToolPackage]::InstalledPackages[$this.PackageId]
```
Duplicate?

ryfu-msft commented on resources/Microsoft.DotNet.Dsc/Microsoft.DotNet.Dsc.psm1:
```diff
@@ -0,0 +1,547 @@
+                    $currentState.Exist = $false
+                    $currentState.Version = $this.Version
+                }
+                else 
+                {
+                    # Don't want to downgrade scenarios
+                    Throw "The package version cannot be less than the installed version. Please provide a version greater than or equal to the installed version."
+                }
```
We shouldn't be throwing on a Get() call. I also believe that we should still support downgrade scenarios even if this is not a common scenario. DSC should be able to set any desired state even if that involves downgrading.

Gijsreyn replied:
Please check once again. The logic is a bit different. Check out the GitHub issue linked at the function. Not all SDKs have the downgrade option ;)

ryfu-msft commented on resources/Microsoft.DotNet.Dsc/Microsoft.DotNet.Dsc.psm1:
```diff
@@ -0,0 +1,547 @@
+    return $resultSet
+}
+
+function Get-SemVer($version)
+{
+    $version -match "^(?<major>\d+)(\.(?<minor>\d+))?(\.(?<patch>\d+))?(\-(?<pre>[0-9A-Za-z\-\.]+))?(\+(?<build>[0-9A-Za-z\-\.]+))?$" | Out-Null
+    $major = [int]$matches['major']
+    $minor = [int]$matches['minor']
+    $patch = [int]$matches['patch']
+    
+    if ($null -eq $matches['pre']) { $pre = @() }
+    else { $pre = $matches['pre'].Split(".") }
+
+    $revision = 0
+    if ($pre.Count -gt 0)
+    {
+        $revision = Get-HighestRevision -InputArray $pre
+    }
+
+    return [version]$version = "$major.$minor.$patch.$revision" 
+}
```
Can consider adding some basic unit tests for this function to verify the behavior

Gijsreyn commented:
@denelon I have also sketch out a proposed solution in the [sdk](https://github.com/dotnet/sdk/issues/44325) repository. I can imagine that it hasn't the highest priority, but if one operation already is exposed, this becoms obsolete. 

If there is anyone whiling to participate, my C# knowledge is rusty. A first set up would greatly assist me.

ryfu-msft commented on resources/Microsoft.DotNet.Dsc/Microsoft.DotNet.Dsc.psm1:
```diff
@@ -262,24 +273,45 @@ function Update-DotNetToolPackage
-        [bool]   $Exist
+        [string] $ToolPathDirectory,
+        [bool]   $Exist,
+        [switch] $Downgrade
```
I don't think a downgrade switch is needed. If the downgrade option is not supported by the dotnet tool, we could just try uninstalling and reinstalling the correct version? I would opt for that option in case a user doesn't know what version is currently on the machine.

Gijsreyn replied:
Got it! Please see latest commit. I have opted to use a helper method ReInstall and just call Uninstall/Install to simplify things. You got 4 scenarios now:

- ReInstall
- Upgrade
- Install
- Uninstall

ryfu-msft commented on resources/Microsoft.DotNet.Dsc/Microsoft.DotNet.Dsc.psm1:
```diff
@@ -382,28 +405,22 @@ class DotNetToolPackage
                 $installedVersion = Get-Semver -Version $currentState.Version
                 $currentVersion = Get-Semver -Version $this.Version
-                if ($currentVersion -gt $installedVersion)
+                if ($currentVersion -gt $installedVersion -or $currentState -le $currentState)
```
```suggestion
                if ($currentVersion -gt $installedVersion -or $currentState -le $currentState)
```

`$currentState -le $currentState`? Not sure what the intent here is but this comparison seems to cover greater than, less than, and equal.

Can you simplify to `$currentVersion -ne $installedVersion`?

Gijsreyn replied:
Great catch Ryan. It has been changed. Tests have successfully run on my side. Please give it a go in the pipelines :)

Would love to see it get to alpha release on the gallery!

ryfu-msft commented:
/azp run

ryfu-msft commented:
/azp run

ryfu-msft reviewed (approved):
Look great! Will work on getting this published to the gallery. Thank you so much for all your hard work on this! 😄

Gijsreyn commented:
And thanks for all the feedback! Please let me know when you get it published

denelon commented:
@Gijsreyn,
It's published: https://www.powershellgallery.com/packages/Microsoft.DotNet.Dsc/0.1.1-alpha

Gijsreyn commented:
@denelon Thanks for letting me know. I'll try to put out some docs tomorrow and spend some love in some upcoming blogs I'm writing. Curious to your thoughts. Cheers.

denelon commented:
@ryfu-msft and I were chatting about the best place to put docs here, and a sample over at https://aka.ms/dsc.yaml inside the Dev Home repository. We could also consider centralizing examples here now that this repository is open to the public. That would allow me to change the redirect to this repository.
